src/checkers/inference/solver/backend/logiqlbackend/LogiQLSerializer.java

```diff
@@ -0,0 +1,276 @@
+import checkers.inference.model.ExistentialConstraint;
+import checkers.inference.model.ExistentialVariableSlot;
+import checkers.inference.model.InequalityConstraint;
+import checkers.inference.model.PreferenceConstraint;
+import checkers.inference.model.RefinementVariableSlot;
+import checkers.inference.model.Serializer;
+import checkers.inference.model.Slot;
+import checkers.inference.model.SubtypeConstraint;
+import checkers.inference.model.VariableSlot;
+import checkers.inference.solver.frontend.Lattice;
+import checkers.inference.solver.util.NameUtils;
+
+/**
+ * LogiQLSerializer converts constraint into string as logiQL data.
+ * 
+ * @author jianchu
+ *
+ */
+public class LogiQLSerializer implements Serializer<String, String> {
+
+    private final Lattice lattice;
+
+    public LogiQLSerializer(Lattice lattice) {
+        this.lattice = lattice;
+    }
+
+    @Override
+    public String serialize(SubtypeConstraint constraint) {
+
+        return new VariableCombos<SubtypeConstraint>() {
+            
+            @Override
+            protected String constant_variable(ConstantSlot subtype, VariableSlot supertype, SubtypeConstraint constraint) {
+                String subtypeName = NameUtils.getSimpleName(subtype.getValue());
+                int supertypeId = supertype.getId();
+                String logiQLData = "+subtypeConstraintLeftConstant(c, v), +constant(c), +hasconstantName[c] = \""
+                        + subtypeName + "\", +variable(v), +hasvariableName[v] = " + supertypeId + ".\n";
+                return logiQLData;
+            }
+
+            @Override
+            protected String variable_constant(VariableSlot subtype, ConstantSlot supertype, SubtypeConstraint constraint) {
+                String supertypeName = NameUtils.getSimpleName(supertype.getValue());
+                int subtypeId = subtype.getId();
+                String logiQLData = "+subtypeConstraintRightConstant(v, c), +variable(v), +hasvariableName[v] = "
+                        + subtypeId
+                        + ", +constant(c), +hasconstantName[c] = \""
+                        + supertypeName
+                        + "\" .\n";
+                return logiQLData;
+            }
+
+            @Override
+            protected String variable_variable(VariableSlot subtype,  VariableSlot supertype, SubtypeConstraint constraint) {
+                String logiQLData = "+subtypeConstraint(v1, v2), +variable(v1), +hasvariableName[v1] = "
+                        + subtype.getId() + ", +variable(v2), +hasvariableName[v2] = "
+                        + supertype.getId() + ".\n";
+                return logiQLData;
+            }
+            
+            @Override
+            protected String constant_constant(ConstantSlot slot1, ConstantSlot slot2, SubtypeConstraint constraint) {
+
+                return defaultAction(slot1, slot2, constraint);
+            }
+
+        }.accept(constraint.getSubtype(), constraint.getSupertype(), constraint);
```

Comment:
Returning a long annonymous class is hard to read. Luqman's changes are easy to understand and maintain. He created inner classes like SubtypeConstraintAtmCombo.